src/compiler/checker.ts

```diff
@@ -6338,6 +6338,10 @@ namespace ts {
                                 // Stop at the first containing function or module declaration
                                 break loop;
                         }
+                        // Preserve old top-level behavior - if the branch is really an empty set, revert to prior type
+                        if (narrowedType === getUnionType(emptyArray)) {

```

Comment:
I respected the note left in the narrowing code about this - if we change how we represent empty unions, then this is the canonical way to find that representation.